pkg/upgrade/upgrade_state.go

```diff
@@ -23,201 +23,76 @@ import (
 	"github.com/go-logr/logr"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/intstr"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	v1alpha1 "github.com/NVIDIA/k8s-operator-libs/api/upgrade/v1alpha1"
+	"github.com/NVIDIA/k8s-operator-libs/api/upgrade/v1alpha1"
 	"github.com/NVIDIA/k8s-operator-libs/pkg/consts"
+	"github.com/NVIDIA/k8s-operator-libs/pkg/upgrade/base"
+	"github.com/NVIDIA/k8s-operator-libs/pkg/upgrade/inplace"
 )
 
-// NodeUpgradeState contains a mapping between a node,
-// the driver POD running on them and the daemon set, controlling this pod
-type NodeUpgradeState struct {
-	Node            *corev1.Node
-	DriverPod       *corev1.Pod
-	DriverDaemonSet *appsv1.DaemonSet
-}
-
-// IsOrphanedPod returns true if Pod is not associated to a DaemonSet
-func (nus *NodeUpgradeState) IsOrphanedPod() bool {
-	return nus.DriverDaemonSet == nil
-}
-
-// ClusterUpgradeState contains a snapshot of the driver upgrade state in the cluster
-// It contains driver upgrade policy and mappings between nodes and their upgrade state
-// Nodes are grouped together with the driver POD running on them and the daemon set, controlling this pod
-// This state is then used as an input for the ClusterUpgradeStateManager
-type ClusterUpgradeState struct {
-	NodeStates map[string][]*NodeUpgradeState
-}
-
-// NewClusterUpgradeState creates an empty ClusterUpgradeState object
-func NewClusterUpgradeState() ClusterUpgradeState {
-	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
-}
-
-// ClusterUpgradeStateManager is an interface for performing cluster upgrades of driver containers
-//
-//nolint:interfacebloat
-type ClusterUpgradeStateManager interface {
+// ExtendedUpgradeStateManager interface purpose is to decouple ApplyState implementation from base package
+// since upgrade pkg is a high level abstaction referencing inplace/requestor (maintenance OP) packages
+type ExtendedUpgradeStateManager interface {
+	// BuildState builds a point-in-time snapshot of the driver upgrade state in the cluster.
+	BuildState(ctx context.Context, namespace string,
```

Comment:
I think that we should add an extension point at a different level. 

a lot of the files in the base package are only used with the inplace mode (cordon manager, drain manager). Because of this, I think that we should move them to the inplace package and move the `ApplyState` method to the ProcessNodeStateManager interface.

That way, we can have two clearly defined different flows for the inplace and requestor modes.